src/EventListener/SessionListener.php

```diff
@@ -67,15 +82,30 @@ public function onKernelResponse(FilterResponseEvent $event)
             return;
         }
 
+        // copied from the Symfony SessionListener, to avoid breaking the lazyness
+        if (!$session = $this->container && $this->container->has('initialized_session')
+                ? $this->container->get('initialized_session')
+                : $event->getRequest()->getSession()
+        ) {
+            return;
+        }
+
+        // Check if this response has a vary header that sounds like it is about the user context
         $varyHeaders = array_map('strtolower', $event->getResponse()->getVary());
         $relevantHeaders = array_merge($this->userIdentifierHeaders, [$this->userHashHeader]);
 
-        // Call default behaviour if it's an irrelevant request for the user context
-        if (0 === count(array_intersect($varyHeaders, $relevantHeaders))) {
-            $this->inner->onKernelResponse($event);
+        if (0 < count(array_intersect($varyHeaders, $relevantHeaders))
+            && ($session->isStarted() || ($session instanceof Session && $session->hasBeenStarted()))
+        ) {
+            $session->save();
+
+            if ($session instanceof Session && $session->hasBeenStarted()) {
```

Comment:
Shouldn't we return always if Vary header check is statisfied? It seems to me that with this, if the check is satisfied, but for anomymous users without session, the original listener still gets called, right?

Reply:
that is on purpose. i want to interfere as little as possible with the default behaviour. the only case where calling the parent after having done `$session->save()` would result in a problem is if the `hasBeenStarted` flag is present - that seems to try to work around the `save` and still detect if a session was open and has been saved...